src/careamics/dataset_ng/dataset/factory.py

```diff
@@ -0,0 +1,404 @@
+    read_kwargs : dict of {str, Any}, optional
+        Additional key-word arguments to pass to the `read_func`.
+    image_stack_loader : ImageStackLoader, optional
+        A function for custom image stack loading. This argument is ignored unless the
+        `data_type` in the `config` is "custom".
+    image_stack_loader_kwargs : {str, Any}, optional
+        Additional key-word arguments to pass to the `image_stack_loader`.
+
+    Returns
+    -------
+    CareamicsDataset[ImageStack]
+        The CAREamicsDataset
+
+    Raises
+    ------
+    ValueError
+        For an unrecognized `data_type` in the `config`.
+    """
+    data_type = SupportedData(config.data_type)
+    dataset_type = determine_dataset_type(
+        data_type, in_memory, read_func, image_stack_loader
+    )
+    if dataset_type == DatasetType.ARRAY:
+        return create_array_dataset(config, mode, inputs, targets)
+    elif dataset_type == DatasetType.IN_MEM_TIFF:
+        return create_tiff_dataset(config, mode, inputs, targets)
+    # TODO: Lazy tiff
+    elif data_type == DatasetType.IN_MEM_CUSTOM_FILE:
+        if read_kwargs is None:
+            read_kwargs = {}
+        assert read_func is not None  # should be true from `determine_dataset_type`
+        return create_custom_file_dataset(
+            config, mode, inputs, targets, read_func=read_func, read_kwargs=read_kwargs
+        )
+    elif dataset_type == DatasetType.CUSTOM_IMAGE_STACK:
+        if image_stack_loader_kwargs is None:
+            image_stack_loader_kwargs = {}
+        assert image_stack_loader is not None  # should be true
+        return create_custom_image_stack_dataset(
+            config, mode, inputs, targets, image_stack_loader, image_stack_loader_kwargs
+        )
+    else:
+        raise ValueError(f"Unrecognized dataset type, {dataset_type}.")
+
+
+def create_array_dataset(
+    config: Union[DataConfig, InferenceConfig],
+    mode: Mode,
+    inputs: Sequence[NDArray[Any]],
+    targets: Optional[Sequence[NDArray[Any]]],
+) -> CareamicsDataset[InMemoryImageStack]:
+    """
+    Create a CAREamicsDataset from array data.
+
+    Parameters
+    ----------
+    config : DataConfig or InferenceConfig
+        The data configuration.
+    mode : Mode
+        Whether to create the dataset in "training", "validation" or "predicting" mode.
+    inputs : Any
+        The input sources to the dataset.
+    targets : Any, optional
+        The target sources to the dataset.
+
+    Returns
+    -------
+    CareamicsDataset[InMemoryImageStack]
+        A CAREamicsDataset
+    """
+    input_extractor = create_array_extractor(source=inputs, axes=config.axes)
+    target_extractor: Optional[PatchExtractor[InMemoryImageStack]]
+    if targets is not None:
+        target_extractor = create_array_extractor(source=targets, axes=config.axes)
+    else:
+        target_extractor = None
+    dataset = CareamicsDataset(config, mode, input_extractor, target_extractor)
```

Comment:
Not important, but could return directly `CAREamicsDataset` without creating a `dataset`